### PR TITLE
Add image_contents data to Go content manifests

### PR DESCRIPTION
**Pull request by athos-ribeiro**
Signed-off-by: Athos Ribeiro <athos@redhat.com>

### CODE REVIEW

**Comment by athos-ribeiro** [comment]
Still missing some unit tests for the ContentManifest class change. @mprahl, I decided not to use the database after all now we can just store the package level dependencies and their dependencies in the DB. The content manifest generation will be type specific though, i.e., unless implemented, empty content manifests will be provided.

**Comment by mprahl** [comment]
@athos-ribeiro could you please move the `ContentManifest` class to different file since it's not going to be a model anymore?

**Comment by athos-ribeiro** [comment]
It is still part of the business logic though, isn't it? Would you rather see it in the  web.utils module or in its own content_manifest module instead?

File: cachito/workers/pkg_managers/gomod.py [hunk 1]
```diff
@@ -187,7 +189,47 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+            parts = [part for part in line.split(" ") if part not in ("")]
+            replaces = None
+
+            name = parts[0]
+            parent_module = parts[1]
+            if len(parts) > 2:
+                version = parts[2]
+            else:
+                # use requested module version
+                version = module_version
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:210]: Why is the version sometimes set and sometimes not? Is this the case where it's the actual package in the module?

If so, please add a comment in the code.

File: cachito/workers/pkg_managers/gomod.py [hunk 2]
```diff
@@ -187,7 +189,47 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+                "from_module": parent_module,
+                "replaces": replaces,
+                "type": "go-package",
+                "version": version,
+            }
+
+            pkg_level_deps.append(pkg)
+
+        # The last item on `go list -deps` is the main package being evaluated
+        package = pkg_level_deps.pop()
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:229]: Could there be more than one package in the module? How is this accounted for? Does it need to be?

File: cachito/workers/tasks/gomod.py [hunk 3]
```diff
@@ -49,4 +50,10 @@ def fetch_gomod_source(request_id, dep_replacements=None):
...
     env_vars.update(config.cachito_default_environment_variables.get("gomod", {}))
     update_request_with_package(request_id, module, env_vars)
-    update_request_with_deps(request_id, module, deps)
+    update_request_with_deps(request_id, module, module_deps)
+
+    # add package deps
+    [p.pop("from_module") for p in [pkg] + pkg_deps]
+    update_request_with_package(request_id, pkg)
+    update_request_with_deps(request_id, pkg, pkg_deps)
+    update_request_with_deps(request_id, pkg, module_deps)
```
Comment (mprahl) [inline cachito/workers/tasks/gomod.py:59]: Rather than associate the `module_deps` with the package, I think it'd make more sense to somehow associate the package with the module. This would reduce duplication in the database for every request and speed up the request as the dependencies wouldn't have to be sent a second time to the PATCH API endpoint, which checks to see if each dependency already exists in the database. This would save a couple minutes to a large module.

**Comment by mprahl** [comment]
> It is still part of the business logic though, isn't it? Would you rather see it in the web.utils module or in its own content_manifest module instead?

I would say a separate module. What might be confusing is that `models.py` are the database ORM models not the models you think of in MVC.

File: cachito/workers/pkg_managers/gomod.py (hunk 2 above)
Comment (athos-ribeiro) [reply cachito/workers/pkg_managers/gomod.py:229]: Yes. The sub-packages would be accounted for in the package level dependencies as the other (external) package level dependenies.

File: cachito/workers/pkg_managers/gomod.py (hunk 1 above)
Comment (athos-ribeiro) [reply cachito/workers/pkg_managers/gomod.py:210]: Yes, these are the packages and sub-packags contained in the current module. I'll include a better comment here

File: cachito/web/models.py [hunk 4]
```diff
@@ -194,6 +252,19 @@ def get_or_create(cls, package):
...
         return package_object
 
+    def to_purl(self):
+        """
+        Generate the PURL representation of the package.
+
+        :return: the PURL string of the Package object
+        :rtype: string
+        """
+        if self.type == "go-package" or "gomod":
```
[the remark this answers is not in the document]
Comment (athos-ribeiro) [reply cachito/web/models.py:262]: Oh, thanks for spotting this one! :)

File: cachito/web/models.py [hunk 5]
```diff
@@ -72,17 +72,75 @@ class ContentManifest:
...
+        """
+        icm = {
+            "metadata": {
+                "icm_version": self.version,
+                "icm_spec": self.json_schema_url,
+                "image_layer_index": self.unknown_layer_index,
+            },
+        }
+        if image_contents:
+            icm["image_contents"] = image_contents
```
[the remark this answers is not in the document]
Comment (athos-ribeiro) [reply cachito/web/models.py:141]: Sounds good. Changing.

File: cachito/workers/pkg_managers/gomod.py [hunk 6]
```diff
@@ -187,7 +189,47 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+        package.pop("replaces")
+
+        return module, module_level_deps, package, pkg_level_deps
```
[the remark this answers is not in the document]
Comment (athos-ribeiro) [reply cachito/workers/pkg_managers/gomod.py:232]: How about a namedtuple?

File: cachito/workers/pkg_managers/gomod.py [hunk 7]
```diff
@@ -187,7 +189,47 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+
+            if len(parts) > 3:
+                # parts[3] is just an arrow "=>" pointing to the replacements
+                replaces = {"type": "gomod", "name": parent_module, "version": version}
+                parent_module = parts[4]
+                version = parts[5]
+
+            pkg = {
+                "name": name,
+                "from_module": parent_module,
```
[the remark this answers is not in the document]
Comment (athos-ribeiro) [reply cachito/workers/pkg_managers/gomod.py:220]: I thought of adding a relation between the module and the package at first, and ended up chanign my mind on the way. I will remove it if we decide not adding the relation in the discussion below.

File: cachito/workers/pkg_managers/gomod.py [hunk 8]
```diff
@@ -187,7 +189,47 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+        # package level dependencies
+        go_list_deps_output = run_gomod_cmd(
+            ("go", "list", "-deps", "-f", "{{if not .Standard}}{{.ImportPath}} {{.Module}}{{end}}"),
+            run_params,
+        )
+
+        pkg_level_deps = []
+        for line in go_list_deps_output.splitlines():
+            # line example: pkg.io/foo/bar pkg.io/foo/var v1.0.0 => pkg.io/foo/bar v1.0.1
+            parts = [part for part in line.split(" ") if part not in ("")]
```
[the remark this answers is not in the document]
Comment (athos-ribeiro) [reply cachito/workers/pkg_managers/gomod.py:201]: This would avoid adding an empty part in case there would be no module info for a package. This case cannot be reached though (at first I thought we would need an additional field that could be null, where this case could be true, hence the reason for this statement). I will remove the statement.

**Comment by athos-ribeiro** [comment]
Thanks for the review, Matt. I addressed most of the comments in my latest push. I will proceed with the rest (first 2 comments) after we decide how to proceed with package/module dependencies. Please, see https://github.com/release-engineering/cachito/pull/187#discussion_r451356141

**Comment by athos-ribeiro** [comment]
Added the new module for content manifests logic and added a dependency between the requested module and the top level package in it.

File: cachito/workers/pkg_managers/gomod.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mprahl) [reply cachito/workers/pkg_managers/gomod.py]: That works, though we use dictionaries in other parts of the code similar to this so I'd prefer consistency but if you have strong feelings towards using a namedtuple, go for it.

**Comment by athos-ribeiro** [comment]
Substituted namedtuples for dicts and added the missing tests

File: cachito/web/content_manifest.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                    "No ICM implementation for '%s' packages", req_dep.package.type
+                )
+
+        # gomod and go-package all refer to the gomod pkg manager. Let's merge their data
+        for k, v in self._gomod_data.items():
+            # v["pkg"] may be None for old DB entries from pre-content-manifest requests
+            if v["pkg"]:
+                self._gopkg_data[v["pkg"]]["sources"].extend(v["sources"])
+
+        top_level_packages = [v for v in self._gopkg_data.values()]
```
Comment (mprahl) [inline cachito/web/content_manifest.py:81]:
```suggestion
        top_level_packages = list(self._gopkg_data.values())
```

File: cachito/workers/pkg_managers/gomod.py [hunk 10]
```diff
@@ -48,9 +48,10 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
     :param dict request: the Cachito request this is for
     :param list dep_replacements: dependency replacements with the keys "name" and "version"; this
         results in a series of `go mod edit -replace` commands
-    :return: a tuple of the Go module itself and the list of dictionaries representing the
-        dependencies
-    :rtype: (dict, list)
+    :return: a namedtuple containing the Go module itself, the list of dictionaries representing the
+        dependencies, the top package level dependency, and a list of dictionaries representing
+        the package level dependencies
+    :rtype: namedtuple
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:54]:
```suggestion
    :rtype: dict
```

File: cachito/workers/pkg_managers/gomod.py [hunk 11]
```diff
@@ -48,9 +48,10 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
     :param dict request: the Cachito request this is for
     :param list dep_replacements: dependency replacements with the keys "name" and "version"; this
         results in a series of `go mod edit -replace` commands
-    :return: a tuple of the Go module itself and the list of dictionaries representing the
-        dependencies
-    :rtype: (dict, list)
+    :return: a namedtuple containing the Go module itself, the list of dictionaries representing the
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:51]: Since you're using a dictionary, could you please describe the keys in here?

File: cachito/workers/pkg_managers/gomod.py [hunk 12]
```diff
@@ -187,7 +189,47 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+            parent_module = parts[1]
+            if len(parts) > 2:
+                version = parts[2]
+            else:
+                # use requested module version
+                version = module_version
+
+            if len(parts) > 3:
+                # parts[3] is just an arrow "=>" pointing to the replacements
+                replaces = {"type": "gomod", "name": parent_module, "version": version}
```
[the remark this answers is not in the document]
Comment (mprahl) [reply cachito/workers/pkg_managers/gomod.py:215]: @athos-ribeiro I'm wondering if we should just not set replaces for Go packages since the package isn't being replaced, only the module is. If you want to know if the Go package was from a replaced module, you should be able to check that from: go package -> go module -> replaces. I may be misunderstanding this though.

[the remark this answers is not in the document]
Comment (mprahl) [reply cachito/workers/pkg_managers/gomod.py:215]: By the way, if you want to keep this away, I'm also okay with that.

File: cachito/workers/pkg_managers/gomod.py [hunk 13]
```diff
@@ -187,7 +189,54 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
...
+        package.pop("replaces")
+
+        gomod = {
+            "module": module,
+            "module_deps": module_level_deps,
+            "pkg": package,
+            "pkg_deps": pkg_level_deps,
+        }
+
+        return gomod
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:239]:
```suggestion
        return {
            "module": module,
            "module_deps": module_level_deps,
            "pkg": package,
            "pkg_deps": pkg_level_deps,
        }
```

File: cachito/workers/pkg_managers/gomod.py [hunk 14]
```diff
@@ -187,7 +189,54 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
             )
             shutil.copytree(tmp_download_cache_dir, str(bundle_dir.gomod_download_dir))
 
-        return module, deps
+        # package level dependencies
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:192]: Could add a log statement here stating that you're getting the package dependencies?

File: cachito/workers/tasks/gomod.py [hunk 15]
```diff
@@ -37,8 +37,9 @@ def fetch_gomod_source(request_id, dep_replacements=None):
 
     log.info("Fetching gomod dependencies for request %d", request_id)
     request = set_request_state(request_id, "in_progress", "Fetching the gomod dependencies")
+    source_path = str(bundle_dir.source_dir)
```
Comment (mprahl) [inline cachito/workers/tasks/gomod.py:40]: This change looks unnecessary

File: cachito/workers/tasks/gomod.py [hunk 16]
```diff
@@ -48,5 +49,8 @@ def fetch_gomod_source(request_id, dep_replacements=None):
         "GOPATH": {"value": "deps/gomod", "kind": "path"},
     }
     env_vars.update(config.cachito_default_environment_variables.get("gomod", {}))
-    update_request_with_package(request_id, module, env_vars)
-    update_request_with_deps(request_id, module, deps)
+    update_request_with_package(request_id, gomod["module"], env_vars)
+    update_request_with_deps(request_id, gomod["module"], gomod["module_deps"] + [gomod["pkg"]])
```
Comment (mprahl) [inline cachito/workers/tasks/gomod.py:53]: The package of the module doesn't seem like a dependency to me. Since the top-level package and the module have the same name, you can correlate the go module to the go package in the "packages" JSON array without this.

Could you please explain your point of view?

File: cachito/web/content_manifest.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    unknown_layer_index = -1
+
+    def __init__(self, request=None):
+        """
+        Initialize ContentManifest.
+
+        :param Request request: the request to generate a ContentManifest for
+        """
+        self.request = request
+        self._packages = None
```
Comment (mprahl) [inline cachito/web/content_manifest.py:22]: This doesn't seem to be used anymore. Could you please remove it?

File: cachito/web/content_manifest.py [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+        if dependency.type == "gomod":
+            icm_source = {"purl": dependency.to_purl()}
+            self._gomod_data[package.id]["sources"].append(icm_source)
+        elif dependency.type == "go-package":
+            if self._gomod_data[package.id]["pkg"]:
+                raise RuntimeError(f'Multiple go-package dependencies for package "{package.id}"')
+            icm_package = dependency.to_purl()
+            self._gomod_data[package.id]["pkg"] = icm_package
+
+    def to_json(self):
```
Comment (mprahl) [inline cachito/web/content_manifest.py:56]: Could you please set `self._gopkg_data` and `self._gomod_data` to empty dictionaries at the start of this method? This will allow the `to_json` method to be run multiple times and get the same results.

This is what currently happens:
```python
>>> manifest = Request.query.get(1).content_manifest
>>> pprint(manifest.to_json())
{'image_contents': [{'dependencies': [{'purl': 'pkg:golang/github.com%2Fpkg%2Ferrors@v0.8.0'},
                                      {'purl': 'pkg:golang/github.com%2Fop%2Fgo-logging@v0.0.0-20160315200505-970db520ece7'},
                                      {'purl': 'pkg:golang/github.com%2FMasterminds%2Fsemver@v1.4.2'},
                                      {'purl': 'pkg:golang/gopkg.in%2Fyaml.v2@v2.2.2'},
                                      {'purl': 'pkg:golang/github.com%2Frelease-engineering%2Fretrodep%2Fv2%2Fretrodep%2Fglide@v2.0.2-0.20190619083402-c5360d4a2f5f'},
                                      {'purl': 'pkg:golang/golang.org%2Fx%2Ftools%2Fgo%2Fvcs@v0.0.0-20190325161752-5a8dccf5b48a'},
                                      {'purl': 'pkg:golang/github.com%2Frelease-engineering%2Fretrodep%2Fv2%2Fretrodep@v2.0.2-0.20190619083402-c5360d4a2f5f'}],
                     'purl': 'pkg:golang/github.com%2Frelease-engineering%2Fretrodep%2Fv2@v2.0.2-0.20190619083402-c5360d4a2f5f',
                     'sources': [{'purl': 'pkg:golang/github.com%2Fpkg%2Ferrors@v0.8.0'},
                                 {'purl': 'pkg:golang/github.com%2FMasterminds%2Fsemver@v1.4.2'},
                                 {'purl': 'pkg:golang/github.com%2Fkr%2Fpretty@v0.1.0'},
                                 {'purl': 'pkg:golang/github.com%2Fkr%2Fpty@v1.1.1'},
                                 {'purl': 'pkg:golang/github.com%2Fkr%2Ftext@v0.1.0'},
                                 {'purl': 'pkg:golang/github.com%2Fop%2Fgo-logging@v0.0.0-20160315200505-970db520ece7'},
                                 {'purl': 'pkg:golang/golang.org%2Fx%2Fcrypto@v0.0.0-20190308221718-c2843e01d9a2'},
                                 {'purl': 'pkg:golang/golang.org%2Fx%2Fnet@v0.0.0-20190311183353-d8887717615a'},
                                 {'purl': 'pkg:golang/golang.org%2Fx%2Fsys@v0.0.0-20190215142949-d0b11bdaac8a'},
                                 {'purl': 'pkg:golang/golang.org%2Fx%2Ftext@v0.3.0'},
                                 {'purl': 'pkg:golang/golang.org%2Fx%2Ftools@v0.0.0-20190325161752-5a8dccf5b48a'},
                                 {'purl': 'pkg:golang/gopkg.in%2Fcheck.v1@v1.0.0-20180628173108-788fd7840127'},
                                 {'purl': 'pkg:golang/gopkg.in%2Fyaml.v2@v2.2.2'}]}],
 'metadata': {'icm_spec': 'https://raw.githubusercontent.com/containerbuildsystem/atomic-reactor/f4abcfdaf8247a6b074f94fa84f3846f82d781c6/atomic_reactor/schemas/content_manifest.json',
              'icm_version': 1,
              'image_layer_index': -1}}
>>> pprint(manifest.to_json())
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/src/cachito/web/content_manifest.py", line 65, in to_json
    self.process_gomod(req_dep.package, req_dep.dependency)
  File "/src/cachito/web/content_manifest.py", line 52, in process_gomod
    raise RuntimeError(f'Multiple go-package dependencies for package "{package.id}"')
RuntimeError: Multiple go-package dependencies for package "1"
```

File: cachito/web/content_manifest.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+                self.process_go_package(req_dep.package, req_dep.dependency)
+
+            else:
+                flask.current_app.logger.debug(
+                    "No ICM implementation for '%s' packages", req_dep.package.type
+                )
+
+        # gomod and go-package all refer to the gomod pkg manager. Let's merge their data
+        for k, v in self._gomod_data.items():
+            # v["pkg"] may be None for old DB entries from pre-content-manifest requests
```
Comment (mprahl) [inline cachito/web/content_manifest.py:77]: Good catch!

File: cachito/web/content_manifest.py [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+            elif req_dep.package.type == "go-package":
+                self.process_go_package(req_dep.package, req_dep.dependency)
+
+            else:
+                flask.current_app.logger.debug(
+                    "No ICM implementation for '%s' packages", req_dep.package.type
+                )
+
+        # gomod and go-package all refer to the gomod pkg manager. Let's merge their data
+        for k, v in self._gomod_data.items():
```
Comment (mprahl) [inline cachito/web/content_manifest.py:76]: Optional: It'd be nice to use more descriptive variable names here.

Comment (mprahl) [inline cachito/web/content_manifest.py:76]: Since you don't use the keys:
```suggestion
        for v in self._gomod_data.values():
```

File: cachito/web/content_manifest.py [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+        self._gomod_data = {}
+
+    def process_go_package(self, package, dependency):
+        """
+        Process go-package package.
+
+        :param Package package: the go-package package to process
+        :param Dependency dependency: the go-package package dependency to process
+        """
+        package = package.to_purl()
```
Comment (mprahl) [inline cachito/web/content_manifest.py:33]: Could you rename this to `package_purl`? It's harder to read when a new value is assigned to the function argument.

File: cachito/web/content_manifest.py [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+
+    def __init__(self, request=None):
+        """
+        Initialize ContentManifest.
+
+        :param Request request: the request to generate a ContentManifest for
+        """
+        self.request = request
+        self._packages = None
+        self._gopkg_data = {}
```
Comment (mprahl) [inline cachito/web/content_manifest.py:23]: Optional: It'd be nice to add comments describing these two properties.

File: cachito/web/content_manifest.py (hunk 21 above)
Comment (athos-ribeiro) [reply cachito/web/content_manifest.py:33]: Changed to purl since it stands for package URL

**Comment by athos-ribeiro** [comment]
Changes:

- Main package no longer depends on main module. We now infer the main package and the main module do have the same name
- No more package replacements: For Cachito, this is tied to the module level dependencies and the semantics should always be inferred from module substitutions when dealing with packages. For instance, a module substitution may introduce or remove packages.
- Addressed remaining comments.

File: cachito/web/content_manifest.py (hunk 19 above)
Comment (athos-ribeiro) [reply cachito/web/content_manifest.py]: We are now dealing with this differently: while we are still not generating ICMs for npm packages, gomod packages in requests without package level dependencies will have ICMs with only the "sources" field populated, i.e., only build time dependencies will be listed for those.

File: cachito/web/content_manifest.py [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+
+    def process_go_package(self, package, dependency):
+        """
+        Process gomod or go-package package.
+
+        :param Package package: the gomod or go-package package to process
+        :param Dependency dependency: the gomod or go-package package dependency to process
+        """
+        purl = package.to_purl()
+        # self._gopkg_data.setdefault(purl, {"purl": purl, "dependencies": [], "sources": []})
```
Comment (mprahl) [inline cachito/web/content_manifest.py:33]: Could you please remove this unnecessary comment?

File: cachito/web/content_manifest.py [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+                purl = package.to_purl()
+                self._gopkg_data.setdefault(purl, {"purl": purl, "dependencies": [], "sources": []})
+            else:
+                flask.current_app.logger.debug(
+                    "No ICM implementation for '%s' packages", package.type
+                )
+
+        for req_dep in self.request.request_dependencies:
+            if req_dep.package.type in ("gomod", "go-package"):
+                self.process_go_package(req_dep.package, req_dep.dependency)
```
Comment (mprahl) [inline cachito/web/content_manifest.py:62]: Optional: Since `process_go_package` is quite small and only used here, it might be easier to read if the code was just included here. I'll leave it up to you though.

File: cachito/workers/pkg_managers/gomod.py [hunk 25]
```diff
@@ -187,7 +190,51 @@ def resolve_gomod(app_source_path, request, dep_replacements=None):
             )
             shutil.copytree(tmp_download_cache_dir, str(bundle_dir.gomod_download_dir))
 
-        return module, deps
+        log.info("Retrieving list of package level dependencies")
```
Comment (mprahl) [inline cachito/workers/pkg_managers/gomod.py:193]:
```suggestion
        log.info("Retrieving the list of package level dependencies")
```

File: tests/test_api_v1.py [hunk 26]
```diff
@@ -1259,6 +1261,118 @@ def test_request_fetch_request_content_manifest_invalid(client, worker_auth_env)
...
+    # emulate worker
+    payload = {"dependencies": sample_deps, "package": sample_package}
+    client.patch("/api/v1/requests/1", json=payload, environ_base=worker_auth_env)
+    payload = {"dependencies": sample_pkg_deps, "package": sample_pkg_lvl_pkg}
+    client.patch("/api/v1/requests/1", json=payload, environ_base=worker_auth_env)
+
+    rv = client.get("/api/v1/requests/1")
+    assert rv.status_code == 200
+    response = rv.json
+    assert "/api/v1/requests/1/content-manifest" in response["content_manifest"]
```
Comment (mprahl) [inline tests/test_api_v1.py:1316]:
```suggestion
    assert response["content_manifest"].endswith("/api/v1/requests/1/content-manifest")
```

File: tests/test_api_v1.py [hunk 27]
```diff
@@ -1259,6 +1261,118 @@ def test_request_fetch_request_content_manifest_invalid(client, worker_auth_env)
...
+
+    rv = client.get("/api/v1/requests/1/content-manifest")
+
```
Comment (mprahl) [inline tests/test_api_v1.py:1319]:
```suggestion
```

File: tests/test_api_v1.py [hunk 28]
```diff
@@ -1259,6 +1261,118 @@ def test_request_fetch_request_content_manifest_invalid(client, worker_auth_env)
...
+    }
+
+    # emulate worker
+    payload = {"dependencies": sample_deps, "package": sample_package}
+    client.patch("/api/v1/requests/1", json=payload, environ_base=worker_auth_env)
+
+    rv = client.get("/api/v1/requests/1")
+    assert rv.status_code == 200
+    response = rv.json
+    assert "/api/v1/requests/1/content-manifest" in response["content_manifest"]
```
Comment (mprahl) [inline tests/test_api_v1.py:1369]:
```suggestion
    assert response["content_manifest"].endswith("/api/v1/requests/1/content-manifest")
```

File: tests/test_api_v1.py [hunk 29]
```diff
@@ -1259,6 +1261,118 @@ def test_request_fetch_request_content_manifest_invalid(client, worker_auth_env)
...
+
+    rv = client.get("/api/v1/requests/1/content-manifest")
+
```
Comment (mprahl) [inline tests/test_api_v1.py:1372]: Optional: Small nitpick :stuck_out_tongue: 
```suggestion
```

**Comment by mprahl** [comment]
@athos-ribeiro could you please add a unit test for making sure the content manifest for a npm request works and is empty?

**Comment by mprahl** [comment]
@athos-ribeiro just a few minor comments. This looks really good. Nice job!

File: cachito/web/content_manifest.py (hunk 24 above)
Comment (athos-ribeiro) [reply cachito/web/content_manifest.py:62]: We will soon have to add code to process npm (and python in the future). I did it this with this in mind so this method won't get too bloated.

File: cachito/web/content_manifest.py (hunk 23 above)
Comment (athos-ribeiro) [reply cachito/web/content_manifest.py]: done

File: cachito/workers/pkg_managers/gomod.py (hunk 25 above)
Comment (athos-ribeiro) [reply cachito/workers/pkg_managers/gomod.py]: applied

**Comment by athos-ribeiro** [comment]
@mprahl I addressed your latest comments and squashed the commits.

There is a unit test covering npm in `tests/test_api_v1.py:1322`. It checks if unsuported package types generate empty an image contents section in the content manifest endpoint.
